core/src/main/scala/kafka/zk/KafkaZkClient.scala

```diff
@@ -1631,26 +1637,88 @@ class KafkaZkClient private (zooKeeperClient: ZooKeeperClient, isSecure: Boolean
       throw KeeperException.create(code)
   }
 
+  private def isZKSessionIdDiffFromCurrentZKSessionId(): Boolean = {
+    zooKeeperClient.sessionId != currentZooKeeperSessionId
+  }
+
+  private def isZKSessionTheEphemeralOwner(ephemeralOwnerId: Long): Boolean = {
+    ephemeralOwnerId == currentZooKeeperSessionId
+  }
+
+  private[zk] def shouldReCreateEphemeralZNode(ephemeralOwnerId: Long): Boolean = {
+    isZKSessionTheEphemeralOwner(ephemeralOwnerId) && isZKSessionIdDiffFromCurrentZKSessionId()
+  }
+
+  private def updateCurrentZKSessionId(newSessionId: Long): Unit = {
+    currentZooKeeperSessionId = newSessionId
+  }
+
   private class CheckedEphemeral(path: String, data: Array[Byte]) extends Logging {
     def create(): Code = {
       val createRequest = CreateRequest(path, data, acls(path), CreateMode.EPHEMERAL)
       val createResponse = retryRequestUntilConnected(createRequest)
       createResponse.resultCode match {
-        case code@ Code.OK => code
+        case code@ Code.OK =>
+          // At this point, we need to save a reference to the zookeeper session id.
+          // This is done here since the Zookeeper session id may not be available at the Object creation time.
+          // This is assuming the 'retryRequestUntilConnected' method got connected and a valid session id is present.
+          // This code is part of the work around done in the KAFKA-7165, once ZOOKEEPER-2985 is complete, this code
+          // must be deleted.
+          updateCurrentZKSessionId(zooKeeperClient.sessionId)
```

Comment:
Actually, we need to update the ZK session id in the case below when getAfterNodeExists(0 returns true. So, perhaps we can save the code for all cases and call updateCurrentZKSessionId() if the saved code is OK.

Reply:
I think you mean when `getAfterNodeExists()` returns `Code.OK`.
If I understood correctly, will be like:

```
var createResultCode = Code.OK
      createResponse.resultCode match {
        case code@ Code.OK =>
          createResultCode = code
        case Code.NODEEXISTS =>
          createResultCode = getAfterNodeExists()
        case code =>
          error(s"Error while creating ephemeral at $path with return code: $code")
          createResultCode = code
      }

      if (createResultCode == Code.OK) {
        // At this point, we need to save a reference to the zookeeper session id.
        // This is done here since the Zookeeper session id may not be available at the Object creation time.
        // This is assuming the 'retryRequestUntilConnected' method got connected and a valid session id is present.
        // This code is part of the work around done in the KAFKA-7165, once ZOOKEEPER-2985 is complete, this code
        // must be deleted.
        updateCurrentZKSessionId(zooKeeperClient.sessionId)
      }

      createResultCode
```

What do you think?

Reply:
Yes, that's exactly what I was thinking.

Reply:
Cool @junrao, I have updated the code as suggested.